Improve trash performance

fjvallarino opened the pull request:
Trash operations are not using Cloud Storage anymore, relying instead on Datastore. A task to handle Cloud Storage deletions was added, plus a task (which will be ran only once) to create all the necessary RvStorageObjects and move all Cloud Storage objects under the --TRASH--/ prefix to root.

@tejohnso please review

tejohnso commented on src/main/java/com/risevision/storage/gcs/BatchMoveToTrash.java:
```diff
@@ -2,160 +2,100 @@
-    storage.objects().delete(bucketName, folderName)
-        .queue(batchDelete, new BatchRenameCallback(Operation.DELETE, folderName));    
+    storageObjects.add(storageObject);
+    removedObjects.add(new RvStorageObject(companyId, objectId));
```
What's going on with these two lines?

fjvallarino replied:
The first one creates a new RvStorageObject to represent the file in Trash. The second removes the RvStorageObject associated with the file in Root. I don't think there is a way of updating the id of a Datastore object.

tejohnso commented on src/main/java/com/risevision/storage/gcs/StorageService.java:
```diff
@@ -141,19 +151,25 @@ public boolean bucketExists(String bucketName) throws ServiceFailedException {
     
     for (StorageObject item : items) {
-      RvStorageObject rvso = datastoreRecords.get(companyId + item.getName());
+      RvStorageObject rvso = datastoreRecords.get(companyId + (isTrash ? Globals.TRASH : "") + item.getName());
```
I see this is what you were talking about earlier.  So the trash string is being used as a boolean to indicate that something is trashed. Is the plan to fix that up eventually so that we have booleans in rvstorageobjects for both trashed and deleted?

fjvallarino replied:
Originally I had a "trashed" attribute (you can see that on previous commits), and a file was in Trash or not depending on that attribute (that step of removing Root's RvStorageObject and creating a new one for Trash was not needed; it was just a matter of updating that field). In fact, RvStorageObject.objectId was not using --TRASH--/ prefix. The problem arose with folders, specifically in the case of removing a nested folder. For instance, if I had "/test" and "/test/sub1" in Root, and deleted "/test/sub1", I would still have "/test" in Root, but I would also have "/test" and "/test/sub1" in Trash. Without adding something to objectId (i.e., --TRASH--/), there was no way to distinguish the "/test" in Root and the one in Trash. So, I went back to the --TRASH--/ prefix. After adding it, I realized I was not needing the "trashed" attribute anymore, and it was not necessary for searching either, since we can search for files in Trash (or not) based on the prefix.

tejohnso replied:
I don't follow.
If a flag was being used, why would there be a "/test" and "/test/sub1" in trash.  Wouldn't there be a bunch of files starting with /test/sub1 that have the trash flag set?

Before trashing test/sub1:
test/1.txt
test/sub1/1.txt

After trashing test/sub1:
test/1.txt
test/sub1/1.txt - with trash flag

fjvallarino replied:
Because I need to have an RvStorageObject for folders too, otherwise I don't know if the folder is in Trash or not.

tejohnso commented:
:+1:

fjvallarino commented:
Added unit testing and updated e2e testing. For the moment, I switched to using the default queue to avoid the issue I was having with e2e test cases on CircleCI.

@tejohnso please review

tejohnso commented:
:+1:
